examples/matching-engine/tests/transaction.rs

```diff
@@ -0,0 +1,263 @@
+            amount
+        );
+    }
+    for (owner, amount) in [
+        (admin_account, None),
+        (owner_a, Some(Amount::ZERO)),
+        (owner_b, None),
+    ] {
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id_a, &matching_chain, owner).await,
+            amount
+        );
+    }
+
+    let mut orders_asks = Vec::new();
+    for price in [4, 2] {
+        let price = Price { price };
+        let order = Order::Insert {
+            owner: owner_b,
+            amount: Amount::from_tokens(4),
+            nature: OrderNature::Ask,
+            price,
+        };
+        let operation = Operation::ExecuteOrder { order };
+        let order_messages = fungible_chain_b
+            .add_block(|block| {
+                block.with_operation(matching_id, operation);
+            })
+            .await;
+
+        assert_eq!(order_messages.len(), 3);
+        orders_asks.extend(order_messages);
+    }
+
+    matching_chain
+        .add_block(|block| {
+            block.with_incoming_messages(orders_asks);
+        })
+        .await;
+
+    fungible_chain_a.handle_received_messages().await;
+    fungible_chain_b.handle_received_messages().await;
+    matching_chain.handle_received_messages().await;
+
+    // Now querying the matching
+    let order_ids_a = get_orders(matching_id, &matching_chain, owner_a)
+        .await
+        .expect("order_ids_a");
+    assert_eq!(order_ids_a.len(), 1);
+    let order_ids_b = get_orders(matching_id, &matching_chain, owner_b)
+        .await
+        .expect("order_ids_b");
+    assert_eq!(order_ids_b.len(), 2);
+
+    // Cancelling the remaining orders
+    let mut orders_cancels = Vec::new();
+    for (owner, order_ids) in [(owner_a, order_ids_a), (owner_b, order_ids_b)] {
+        for order_id in order_ids {
+            let order = Order::Cancel { owner, order_id };
+            let operation = Operation::ExecuteOrder { order };
+            let order_messages = fungible_chain_a
+                .add_block(|block| {
+                    block.with_operation(matching_id, operation);
+                })
+                .await;
+            assert_eq!(order_messages.len(), 2);
+            orders_cancels.extend(order_messages);
+        }
+    }
+
+    matching_chain
+        .add_block(|block| {
+            block.with_incoming_messages(orders_cancels);
+        })
+        .await;
+
+    fungible_chain_a.handle_received_messages().await;
+    fungible_chain_b.handle_received_messages().await;
+
+    // Check balances
+    for (owner, amount) in [
+        (owner_a, Amount::from_tokens(3)),
+        (owner_b, Amount::from_tokens(6)),
+    ] {
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id_a, &matching_chain, owner).await,
+            Some(amount)
+        );
+    }
+    for (owner, amount) in [
+        (owner_a, Amount::from_tokens(3)),
+        (owner_b, Amount::from_tokens(5)),
```

Comment:
Same for a missing B.

Reply:
I added an additional explanation about that and tests.